### DISCUSSION TITLE
Audio formats for audiobooks.

### DISCUSSION THREAD

**Discussion by atomotic** [readium/go-toolkit#78] [Ideas]
I am starting a new discussion dedicated to audio formats for audiobooks. In previous community calls, we talked about the importance of making format conversion operations available in go-toolkit.

However, before we start coding, **it is essential to create a matrix table of the formats we wish to support**. This step is crucial because adding dependencies may increase the complexity unnecessarily.

We need to analyze format support in two steps:

# Input conversion or recoding

Given an origin directory containing audio files in format `{source}` we may want to :
* convert them in format `{destination}` into a temporary directory before packaging them
* recoding (changing bitrate etc) and saving new files into a temporary directory before packaging the

The tool for the job is ffmpeg, [ffmpeg-go](https://github.com/u2takey/ffmpeg-go) is quite simple to use. It's just a wrapper around the ffmpeg binary, which should be available in the os running the cli. I wrote about it here:
https://github.com/readium/go-toolkit/discussions/71#discussioncomment-5239063

# Manifest generation
When generating Readium Manifest we need to extract two properties of the audio file: the *mimetype* and the *duration*

## mimetype
[mimetype](https://github.com/gabriel-vasile/mimetype) can be used for extracting the mimetype

## duration

To identify the duration of multiple file formats, we have two choiches:

### 1. Use a decoder library

Until now, I've only supported mp3 and used and I've used [go-mp3](github.com/hajimehoshi/go-mp3) in this way:

```
audiofile, _ := os.Open(file)
d, _ := mp3.NewDecoder(audiofile)
const sampleSize = 4
samples := d.Length() / sampleSize
audioLength := samples / int64(d.SampleRate())
duration = float64(audioLength)
```

What other format should we support? We need to use a decoder library for each format and we should prefer a pure Go implementation. Here are some libraries we can use:

* [flac](https://github.com/mewkiz/flac)
* [aac](https://github.com/gen2brain/aac-go)
* [opus](https://github.com/pion/opus)

### 2. Use ffprobe

```
~ ffprobe -show_entries format=filename,size,duration -of json -i natureofacrime_00_conrad.opus 

{
Input #0, ogg, from 'natureofacrime_00_conrad.opus':
  Duration: 00:10:59.42, start: 0.000000, bitrate: 31 kb/s
  Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
    Metadata:
      encoder         : Lavc59.37.100 libopus
      title           : 00 - Preface
      artist          : Joseph Conrad
      track           : 1
      album           : The Nature of a Crime
      genre           : speech
      TLEN            : 659.46
      comment         : https://archive.org/details/nature_crime_ml_2302_librivox
    "format": {
        "filename": "natureofacrime_00_conrad.opus",
        "duration": "659.421458",
        "size": "2628238"
    }
}
```

[go-ffprobe](https://github.com/vansante/go-ffprobe) can be used, but likewise *go-ffmpeg* needs the binary available in the OS.

**Comment by mickael-menu** [comment]
Regarding mimetype detection, we already have our own utility in the toolkit which should be preferred: https://github.com/readium/go-toolkit/blob/main/pkg/mediatype/mediatype.go

If really needed, we might extend it using third-party dependencies like [`mimetype`](https://github.com/gabriel-vasile/mimetype).
